Te 4.1

mojiiba opened the pull request:
This pull refactors the TE 4.1 test to use the gribi helper APIs.

mojiiba commented:
@liulk Could you please review this PR?

liulk reviewed (changes requested):
Hey sorry I dropped the ball. After spending some time looking into this refactoring, I agree that it made the code much easier to follow. 👍

mojiiba commented on feature/gribi/ate_tests/base_leader_election_test/base_leader_election_test.go:
```diff
@@ -182,201 +180,68 @@ func testTraffic(t *testing.T, ate *ondatra.ATEDevice, top *ondatra.ATETopology,
-	// Configure IPv4 route for 198.51.100.0/24 pointing to ATE port-3 via clientB.
-	configureIPv4ViaClientB(t, args)
+	// Configure IPv4 route for 198.51.100.0/24 pointing to ATE port-3 via clientB as the leader.
+	args.clientB.BecomeLeader(t)
```
Any assumption based on the election id can makes the test case flaky since the client can only be leader when it uses an election id higher than the last known election id (can be any number) on the server side, i.e., there is no guarantee that client can be master  by setting election id to 11. So, I suggest to modify the test case description accordingly by using the last know election id +1 (BecomeLeader does this) rather than a fixed number.

liulk replied:
I see what you're saying. I can see that the test can try to work around some of the problems about accumulating device states from previous runs, but ultimately I think our static binding and other Ondatra bindings should be able to consistently reset the device to a known state between test runs.

The reservation system that we use internally already does that, but this hasn't been implemented in the static binding. Most of the test plans were written assuming a clean state, and I think that's not only for clarity but also for a pragmatic reason. Some of the states are finite resources, such as available LAG IDs or VLAN IDs that run out quickly after repeated test runs.

@greg-dennis and I will be working something out about device state reset. I think for now, we can do BecomeLeader here, but let's mark it as a TODO in the test plan to revert back to hard-coded election IDs. Floating election IDs could be annoying to debug when a test is run on a non-compliant device.

liulk replied:
By the way, is there a way to reset gRIBI election ID on Cisco?

mojiiba replied:
There is no normal procedure to do that. We probably need to reset the gribi process. @mingyangcisco  please correct me if I am wrong?

liulk replied:
Continue discussion in #147

mojiiba commented on feature/gribi/ate_tests/base_leader_election_test/base_leader_election_test.go:
```diff
@@ -182,201 +180,68 @@ func testTraffic(t *testing.T, ate *ondatra.ATEDevice, top *ondatra.ATETopology,
-	configureIPv4ViaClientB(t, args)
+	// Configure IPv4 route for 198.51.100.0/24 pointing to ATE port-3 via clientB as the leader.
+	args.clientB.BecomeLeader(t)
+	configureIPv4ViaLeaderClientInstalled(t, args, args.clientB, atePort3.IPv4)
```
sure

mojiiba commented on feature/gribi/ate_tests/base_leader_election_test/base_leader_election_test.go:
```diff
@@ -182,201 +180,68 @@ func testTraffic(t *testing.T, ate *ondatra.ATEDevice, top *ondatra.ATETopology,
-	args.clientA.Modify().UpdateElectionID(t, 12, 0)
+	// Configure IPv4 route for 198.51.100.0/24 pointing to ATE port-3 via clientA without making it the leader.
+	// The entry should not be installed since clientA is not the leader.
+	configureIPv4ViaNonLeaderClient(t, args, args.clientA)
```
sure

mojiiba commented on feature/gribi/ate_tests/base_leader_election_test/base_leader_election_test.go:
```diff
@@ -182,201 +180,68 @@ func testTraffic(t *testing.T, ate *ondatra.ATEDevice, top *ondatra.ATETopology,
 func testIPv4LeaderActiveChange(ctx context.Context, t *testing.T, args *testArgs) {
-	// Configure IPv4 route for 198.51.100.0/24 pointing to ATE port-3 via clientB.
-	configureIPv4ViaClientB(t, args)
+	// Configure IPv4 route for 198.51.100.0/24 pointing to ATE port-3 via clientB as the leader.
```
sure

mojiiba commented on feature/gribi/ate_tests/base_leader_election_test/base_leader_election_test.go:
```diff
@@ -182,201 +180,68 @@ func testTraffic(t *testing.T, ate *ondatra.ATEDevice, top *ondatra.ATETopology,
 
-	// Configure IPv4 route for 198.51.100.0/24 pointing to ATE port-2 via clientA with election ID of 12.
-	configureIPv4ViaClientAInstalled(t, args)
+	// Configure IPv4 route for 198.51.100.0/24 pointing to ATE port-2 via clientA that is the active leader.
```
sure

mojiiba commented on feature/gribi/ate_tests/base_leader_election_test/base_leader_election_test.go:
```diff
@@ -182,201 +180,68 @@ func testTraffic(t *testing.T, ate *ondatra.ATEDevice, top *ondatra.ATETopology,
-
-	// Modify the election ID of client A to 12 so clientA becomes the active Leader.
-	args.clientA.Modify().UpdateElectionID(t, 12, 0)
+	// Configure IPv4 route for 198.51.100.0/24 pointing to ATE port-3 via clientA without making it the leader.
```
sure

mojiiba commented on feature/gribi/ate_tests/base_leader_election_test/base_leader_election_test.go:
```diff
@@ -404,56 +267,60 @@ func TestElectionIDChange(t *testing.T) {
-
-		tt.fn(ctx, t, args)
-	})
+	for _, tt := range tests {
```
sure

mojiiba commented on feature/gribi/ate_tests/base_leader_election_test/base_leader_election_test.go:
```diff
@@ -404,56 +267,60 @@ func TestElectionIDChange(t *testing.T) {
-		name string
-		desc string
-		fn   func(ctx context.Context, t *testing.T, args *testArgs)
+	tests := []struct {
```
sure

mojiiba commented on feature/gribi/ate_tests/base_leader_election_test/base_leader_election_test.go:
```diff
@@ -404,56 +267,60 @@ func TestElectionIDChange(t *testing.T) {
+				Persistence: tt.persistance,
+			}
+			defer clientA.Close(t)
+			err := clientA.Start(t)
```
sure

mojiiba commented:
@liulk I have addressed your comments. Also, I changed the test code to match the description for using a specific election ids. However, I think gribi provides mechanisms to execute the test cases in a clean state using flush and election id update without reloading the router, so I will open an issue to discuss this and then make the necessary changes if required.

liulk commented:
/gcbrun
